### ISSUE TITLE
Übersicht Wishes

### ISSUE THREAD

**Issue by hawkrives** [felixhageloh/uebersicht#123]
*Note: Please feel free to close this issue. It's a bunch of stuff I'd like to work on for Übersicht, but haven't ever gotten around to. I'm just putting them out here to get some, er, initial feedback on their viability.*

I have several things that I would like to do with Übersicht, in increasing levels of viability/intrusiveness. However, I believe that all of these changes could be made in a completely backwards-compatible fashion, so that *all* existing widgets would continue to work. 

I've had this document floating around since about October 2014.

## Simplify the build system
Grunt is overkill for what Übersicht does with it. I would like to remove grunt entirely, and replace it with npm scripts. 

`package.json`
```json
"config": {
  "browserify": "-t coffeeify --no-detect-globals"
}
"scripts": {
  "jasmine:fs": "npm test",
  "spec": "npm run jasmine:fs && npm run browserify:specs && npm run jasmine",
  "spec:frontend": "npm run browserify:specs && npm run jasmine:src",
  "browserify:server": "browserify -e server.coffee -o release/server.js $npm_package_config_browserify {-i\  {fs,path,child_process,connect,coffee-script,stylus,nib,fsevents}}",
  "browserify:client": "browserify -e client.coffee -o release/public/main.js $npm_package_config_browserify {-i\  {connect,fs,path,child_process,fsevents,coffee-script,stylus,nib,minimist}}",
  "browserify:specs": "browserify -e specs/frontend/**/*_spec.coffee -o release/public/main.js $npm_package_config_browserify {-i\  {stylus,nib}}",
  "jasmine:src": "jasmine JASMINE_CONFIG_PATH=./specs/jasmine.json",
  "release": "npm run browserify:server & npm run browserify:client & wait",
  "default": "npm run spec && npm run release && npm run watch",
  "start": "coffee server.coffee",
  "test": "jasmine-node --coffee --captureExceptions --forceexit spec/backend",
  "watch": "nodemon --ext=styl,coffee -w src/css/**/*.styl -w src/**/*.coffee -w client.coffee -w server.coffee -w spec/**/*_spec.coffee --exec 'npm run spec && npm run release'"
},
"devDependencies": {
	"jasmine": "",
	"nodemon": "",
}
```

`specs/jasmine.json` (for [`jasmine`](https://www.npmjs.com/package/jasmine))
```json
{
  "spec_dir": "specs",
  "spec_files": [
    "frontend_specs.js"
  ],
  "helpers": [
    "vendor/*.js"
  ]
}
```

They can be run by `npm run <name>`, or, in the case of `test`, `serve`, and any others that shadow the name of an `npm` command, by `npm <command>`.

Instead of `grunt release`, you'd run `npm run release`.

## Update dependencies
- Sparkle has gotten some new releases recently – it's aliiive! (https://github.com/sparkle-project/Sparkle/releases).
- Coffeescript has had *two* point releases since the version in `package.json`

## Pin dependencies
Speaking of the `package.json`.

I've run into several points of trouble while trying to hack on Übersicht, usually relating to semver ranges. In my opinion, they're great for libraries, but not so great for applications—it's too hard to ensure that everyone has the same version of things, which is a real problem when debugging.

I'd like to specify absolute versions of dependencies / devDependencies in `package.json`, and use `npm shrinkwrap` to pin everything down. That way, when I change something, I won't be running into spurious compiler output changes ([`<cough>`pull 122`</cough>`](https://github.com/felixhageloh/uebersicht/pull/122#issuecomment-129936238)).

## ES2015 (neè ES6)
I really (really) like writing things in ES6. Of course, Safari doesn't support es6, but if you look at [kangax's tables][es6-compat], there's a column labeled "Babel". Babel can be run as part of a node server, and would allow Übersicht to run widgets written in ES6 as easily as it can run coffeescript widgets today. 

See [pull #122](https://github.com/felixhageloh/uebersicht/pull/122#issuecomment-129936238).

## Modules
The current widget system of being a JS object-minus-braces is nice, but I'd prefer to just be able to `export` the things that Übersicht requires and let the rest live in a file somewhere. Maybe run the files through browserify as part of the server process?

I've been frustrated several times at not being able to declare variables at the top level of my JS widgets; bundling (and named exports) should solve that.

ES6 specifies the `export` keyword, to explicitly name exports, so all we'd need to do is change the bundling process—the "grab attributes from an object" code would continue to work exactly like it does now. 

Before: (should continue to work!)

```coffee
refreshFrequency: '1s'
render: () -> "<p>Hi there</p>"
```

After: (coffeescript)

```coffee
module.exports.refreshFrequency: '1s'
module.exports.render: () -> "<p>Hi there</p>"
```

After: (es2015)

```coffee
export let refreshFrequency = '1s'
export let render = () => "<p>Hi there</p>"
```

## React
As you mentioned in [issue #114](https://github.com/felixhageloh/uebersicht/issues/114#issuecomment-118026742), React is nice. 

I'd like to rebuild the widget system on top of react and webpack.

We could either let widgets opt-in to react support, via a file extension or something (`.react.js`?), or, with Babel support, just check what `render` returns.

We could probably use React's `dangerouslySetInnerHTML` to continue running existing widgets within React components.

I read an interesting article recently about how to prevent a runtime error in user-supplied components from breaking the entire page; it seems like a [nice enough solution](https://nylas.com/blog/react-plugins)

## Conclusion

So. That's what I've got.
If I were to work on any of these topics, would you perchance accept a pull request?
If you had your druthers, which (if any) would you prefer me to work on?

**Comment by felixhageloh** [comment]
good list - many of the issues are on my radar as well. What I'm not sure about is whether to release them as updates or start with Übersicht 2. Obviously, if we would switch to something like react, widgets wouldn't be compatible anymore - which is why a v2 would make sense

**Comment by chrishoage** [comment]
+1 for React and Named exports. It really bugs me that the widgets are not valid JS. https://gist.github.com/felixhageloh/34645a899a0f22f583bb will not lint.

**Comment by hawkrives** [comment]
@felixhageloh After a bunch more thinking, I must agree that a v2 would be necessary.

(Then I thought about possibilities for opting-in to a new format, but then it would have to include two widget systems, and – no. I've abandoned that line of thought.)

**Comment by thousand** [comment]
I definitely think ES2015/ES6 is the way to go. It would also be helpful to allow inclusion of modules for breaking down complex widgets into more than one file.

**Comment by alexisvincent** [comment]
@felixhageloh Theres so many awesome widgets I'd like to bring to this project, but all rely on es6 and being able to import external node libs. This project would be even more amazing then it already is if only it allowed full support for node modules... Even supporting plain require would make this ok :)

**Comment by felixhageloh** [comment]
@alexisvincent about to release CommonJS (plain require) support!
I also plan to release ES2015 widgets with a virtual DOM soon after that

**Comment by felixhageloh** [comment]
beta testers wanted btw!

**Comment by alexisvincent** [comment]
Woohoo! Really happy to beta test. How do I get a copy?

**Comment by felixhageloh** [comment]
here you go: https://www.dropbox.com/s/qq1oc4eohqw5frh/Uebersicht-0.9beta.app.zip

**Comment by alexisvincent** [comment]
@felixhageloh Thanks! Do you have an example commonjs module?

**Comment by felixhageloh** [comment]
well ... anything that can be installed with `npm install` (and works in the browser). I'll show an example using this module https://www.npmjs.com/package/date-math.

Create a folder called `date-example` for you new widget and then in the terminal do:
```sh
cd /path/to/widget/dir/date-example/
npm install date-math
```

You can then write a widget like this (`date-example/index.coffee`):

```coffeescript
Dates = require('date-math')
myBirthday = new Date('1981-11-21')

refreshFrequency: '1d'

command: (callback) ->
  now = new Date()
  days =  Dates.day.diff(myBirthday, now)
  callback(null, days)

style: "color: #fff"

render: (days) ->
  "<h1>You were born #{days} days ago</h1>"
```

Note that I'm making use of another feature as well: commands can now also be plain JS functions!

**Comment by julienvincent** [comment]
@felixhageloh Can this be done with a root js file?

**Comment by alexisvincent** [comment]
@felixhageloh Sorry, I assumed you could do this
```js
module.exports.render = function(output) {
    return '<p>lol</p>';
}
```

**Comment by alexisvincent** [comment]
Whats cool with cjs support now is that widget writers can use babel/require if they so please

**Comment by felixhageloh** [comment]
@alexisvincent that will be the case for the new widget es2015 widget syntax (it is included in this release btw, but it is unstable still). It will look something like this: 

```js
const style = {
  color: '#fff',
  left: 0
};

export const command = 'top';
export const refreshFrequency = '1s';

export function render(output, dispatch) {
  return <div style={style}>
    {output}
  </div>;
}
```

**Comment by felixhageloh** [comment]
@julienvincent I will need to double check again. The current widget format makes standard JS syntax hard to work with. However, as you can see in the comment above, the future format for writing widgets will use JS (es2015)

**Comment by alexisvincent** [comment]
Cool. We're having issues with nested files that are being parsed as well. How do we stop this from happening? Hidden files/folders?

**Comment by felixhageloh** [comment]
good point - all files inside `/node_modules`, `/lib` and `/src` are now ignored!

**Comment by alexisvincent** [comment]
@felixhageloh Awesome :D

**Comment by alexisvincent** [comment]
Can you publish a copy with /node_modules, /lib and /src ignored?

**Comment by felixhageloh** [comment]
the beta I've linked should do that! Unless you discovered the first bug :)

**Comment by alexisvincent** [comment]
hmm. Doesn't seem to be working... But that may just be the following bug breaking things

adding the following index.js file breaks it
```js
    render: require('./node_modules/background.js').render,
    update: require('./node_modules/background.js').update,
    command: require('./node_modules/background.js').command,
    style: require('./node_modules/background.js').style,
    refreshFrequency: require('./node_modules/background.js').refreshFrequency,
    afterRender: require('./node_modules/background.js').afterRender,
```

**Comment by felixhageloh** [comment]
what's the error? Does background.js show up as a separate widget?

**Comment by felixhageloh** [comment]
btw, if you put it in `/node_modules` the file will be available as just `require('background')`. That might well be the problem. I'd suggest to not use node_modules for your own files. It is usually just used for anything you install with `npm install`

**Comment by alexisvincent** [comment]
Theres two separate issues. The lib/ one might not be an issue. My problem is that an index file like the above breaks ubersight. Yeah I know I was just doing it because I didn't think lib/ was ignored

**Comment by alexisvincent** [comment]
I'll give a more detailed error in a moment. Just testing something

**Comment by alexisvincent** [comment]
Ok I specifically get the following error 
exports — client.js:4507/Users/alexisvincent/Library/Application Support/Übersicht/widgets/background/index.js: Error: Line 1: Unexpected identifier while parsing file: /Users/alexisvincent/Library/Application Support/Übersicht/widgets/background/lib/background.js

with the following two files:
index.js
```js
render: require('./lib/background.js').render,
// update: require('./lib/background.js').update,
// command: require('./lib/background.js').command,
// style: require('./lib/background.js').style,
// refreshFrequency: require('./lib/background.js').refreshFrequency,
// afterRender: require('./lib/background.js').afterRender,
```

./lib/background.js
```js
function render() {
    return "<p>lol</p>";
}

module.exports.render = render;
```

**Comment by felixhageloh** [comment]
just realized that are using JS for you index file. I'll need to double check if I can make CommonJS work in this case, but for now it only works for coffeescript widgets (any module you require can be written in JS tho). 

The reason for this is mainly that the current format for JS widgets is actually invalid javascript. So I need some 'tricks' to make it work which is hard to combine with CommonJS support. It should be possible of course, but since I'm planing to roll out the new widget format soon anyway, I decided not to spend to much time to also port CommonJS support to plain JS widgets. I'm just making sure that any existing widgets don't break

**Comment by alexisvincent** [comment]
@felixhageloh Cool got it working. But the web engine is damn slow. Tried rendering a particle canvas, but frame rate was so so bad. Tried loading http://127.0.0.1:41416 in safari directly and it was butter smooth. Are you adding restrictions to the web view?

**Comment by alexisvincent** [comment]
I don't think you're using the new WKWebView

**Comment by felixhageloh** [comment]
I'm not adding any restrictions to the web view, but it is known to be slower than Safari. I'm also still on an older version and apparently the latest one (WKWebView) is closer to Safari in performance. 

I'm hoping to upgrade to WKWebView but I haven't figured out how to show the debug console programmatically (any pointers welcome btw!)

**Comment by alexisvincent** [comment]
@felixhageloh I'm not sure how to do this either... Have had limited experience with Objective C

**Comment by alexisvincent** [comment]
http://stackoverflow.com/questions/25200116/how-to-show-the-inspector-within-your-wkwebview-based-desktop-app

**Comment by felixhageloh** [comment]
Thanks! 
I was aware of this and it is part of the solution: It enables the inspector but you still need to right click on the web view itself to open the inspector. I was hoping to keep the current behavior were you can open it directly from the Übersicht menu

**Comment by alexisvincent** [comment]
I see, I don't know if this will help? Seems to be a private API that was added https://bugs.webkit.org/show_bug.cgi?id=135811

**Comment by julienvincent** [comment]
@felixhageloh I think that the performance gained from WKWebView is still worth sacrificing that feature. In the case where the developer wants access to the terminal - They can use their web browser and directly access the server.

**Comment by alexisvincent** [comment]
@felixhageloh I agree

**Comment by felixhageloh** [comment]
@alexisvincent thanks again :) From what I can tell from the patch, tho, it also just exposes a hook to enable the inspector, but not to show it. But I think the answer does lie somewhere close, i.e. by digging through the webkit repo. I did have a look a while back and found some interesting bits, but no time to follow up on it yet

@julienvincent I agree in principle and I would also be willing to sacrifice some features once I start working on this (upgrading to WKWebView). Having said that, I think that for the majority of users/widgets out there performance is not that important, but ease of use is. So that's always something I need to consider when working on new features/improvements.

**Comment by alexisvincent** [comment]
@felixhageloh I see, I must just voice that while this project is awesome, for anything more exciting then stock static reporting the standard webview is unusable. I have 2 really awesome widgets I'd really like to submit but the performance is a dead killer.

**Comment by felixhageloh** [comment]
@alexisvincent I'm pretty sure now I will look into upgrading to WKWebView after I release the new widget syntax. Curious to see your widgets btw!

**Comment by alexisvincent** [comment]
@felixhageloh cool :) I'd be happy to send you videos.

**Comment by hawkrives** [comment]
I think that all of my wishes have come true! 😭 

- [x] Build system – no more grunt
- [x] es2015 – babel ftw!
- [x] react/vdom (well, jsx) – snabbdom!

About the only thing I see left are docs (I'll help) and style exports from jsx widgets, since `VirtualDomWidget.js` doesn't do anything with the Stylus stylesheets.

I should probably open a new issue for the style exports, at least.

**Comment by hawkrives** [comment]
(also, congrats on the vdom integration, @felixhageloh! i'm really liking it.)

**Comment by felixhageloh** [comment]
thanks @hawkrives. I'm still unsure how to release the new widget syntax as the default. In the long run I would prefer not having to support two (even 3 if you count plain js) syntaxes anymore, so I'm considering to release it as Übersicht 2, which will have its own new widget gallery. 
Btw can you elaborate what you mean with style exports for jsx widgets? 

@alexisvincent @julienvincent I forgot to mention it here, but since 1.0 Übersicht does use WKWebView (and I managed to get the inspector working there as well)

**Comment by julienvincent** [comment]
@felixhageloh awesome! 👍

**Comment by vkravets** [comment]
@felixhageloh BTW where I can find widgets with new syntax?

**Comment by hawkrives** [comment]
@vkravets I made a copy of a Reminders widget tonight in the new format, if you want it: https://gist.github.com/hawkrives/9097548e6536bb9c91034849226088f2. I referenced [snabbdom-jsx](https://github.com/yelouafi/snabbdom-jsx) quite a bit while making it.

@felixhageloh With the classic widgets, the `style` string was interpreted as a Stylus spreadsheet, with all the rules prefixed with the widget id. I can't figure out how to replicate that process with the JSX widgets; nothing seems to care about styles there. My best solution so far has been to do inline styles with the `style` prop, but that's kinda awkward.

**Comment by felixhageloh** [comment]
@vkravets here is another example of the new syntax https://gist.github.com/felixhageloh/94d5dd41cc5ff8a633900ffb28bd195a

@hawkrives I see! No the idea is indeed to use the style prop. It prevents you from having to write css/stylus inside a string, which always felt a bit awkward. It also avoids all the 'leakiness' and side effects that CSS brings

**Comment by vkravets** [comment]
Cool, thanks a lot! =)

**Comment by hawkrives** [comment]
@felixhageloh … grumble 😜 

I've got to say, using objects for styling purposes is my least favorite part of working with [React Native][stylesheet], and it's why I love [CSS Modules][css-modules] so much – because I really like (or maybe I've just gotten used to) CSS syntax. Maybe I'm just stuck in my ways or have a mild case of Stockholm Syndrome, but … bleh.

My concerns are partially alleviated by the snabbdom styles plugin that you've loaded, so I don't have to quote my hyphenated rules 😄 

I'll live, and most people probably won't even care, especially when bundled with the new widget format as a single breaking change. 

[stylesheet]: http://facebook.github.io/react-native/releases/0.33/docs/stylesheet.html#stylesheet
[css-modules]: https://github.com/css-modules/css-modules

(BTW, thank you so much for your work on Übersicht over the years. It's been indispensable for one of my projects, and I enjoy having on my background as well.)

**Comment by felixhageloh** [comment]
@hawkrives ok noted :D.
CSS Modules solve the leakiness problem but I kind of like the simplicity of having just one thing, i.e. JavaScript, to deal with (also from an implementation point of view it is a lot easier because I only need to load common js modules). I can see how writing CSS can be nicer syntax wise tho - and of course you don't have :hover etc.

Btw, as a tip from looking at your widget (great job btw): You can also break up your widget into smaller components, each including their own styles. I find that it becomes a lot more manageable this way. A good rule of thumb is to only have a single JSX render function per component.

**Comment by hawkrives** [comment]
@felixhageloh yeah, I kinda did that with the separate functions – I haven't internalized the fact that I can require other files yet :D. I'm going to need to experiment with what level of modularity I want to use for composing styles together, too.

I didn't make the widget, just ported it to JSX from [RemindersWidget](http://tracesof.net/uebersicht-widgets/#RemindersWidget) – credit where credit's due and all that :).

(I already found myself missing `::after` and friends, but I don't think anyone's really come up with a replacement for that without CSS. I wonder if Safari 10's Shadow DOM implementation limits the scope of styles?)

**Comment by hawkrives** [comment]
Yes! Safari 10's bringing [Shadow DOM support](https://developer.apple.com/library/prerelease/content/releasenotes/General/WhatsNewInSafari/Articles/Safari_10_0.html#//apple_ref/doc/uid/TP40014305-CH11-DontLinkElementID_5), and from skimming their [introductory post](https://webkit.org/blog/4096/introducing-shadow-dom-api/) it seems like it scopes styles!

I wonder if anyone's tried running Snabbdom in a shadow dom? Virtual DOM inside a shadow DOM?

… I'm really excited about this, but I'm not sure (a) what it'll let Übersicht do, truthfully, or (b) how to support it on Sarai 9 with doesn't have shadow dom… thoughts for another day.

**Comment by felixhageloh** [comment]
intersting! That means on macOS Sierra you should be able to play around with it already

**Comment by hawkrives** [comment]
… adding to my list of things to research: CSS Variables, and the possibility of using them for global styles / themes in widgets.
